src/server/page_server.cpp

```diff
@@ -204,18 +204,80 @@ page_server::connection_handler::receive_start_catch_up (tran_server_conn_t::seq
   unpacker.unpack_int (port);
   cublog::lsa_utils::unpack (unpacker, catchup_lsa);
 
-  er_log_debug (ARG_FILE_LINE, "receive_start_catch_up: hostname = %s, port = %d, LSA = (%lld|%d)\n", host.c_str (), port,
-		LSA_AS_ARGS (&catchup_lsa));
+  _er_log_debug (ARG_FILE_LINE, "[CATCH-UP] Start catching up with the follower (%s:%d), until LSA = (%lld|%d)\n",
+		 host.c_str (), port,LSA_AS_ARGS (&catchup_lsa));
   if (port == -1)
     {
       return; // TODO: It means that the ATS is booting up, it will be set properly after ATS recovery is implemented.
     }
 
   // TODO: A thread will take the catch-up including establishing connection to avoid blocking ATS->PS reqeusts.
-  m_ps.connect_to_followee_page_server (std::move (host), port);
+  // All the below here will be taken up.
+  {
+    m_ps.connect_to_followee_page_server (std::move (host), port);
+
+    assert (!log_Gl.hdr.append_lsa.is_null ());
+    assert (!catchup_lsa.is_null ());
+    assert (catchup_lsa >= log_Gl.hdr.append_lsa);
+
+    if (log_Gl.hdr.append_lsa == catchup_lsa)
+      {
+	_er_log_debug (ARG_FILE_LINE, "[CATCH-UP] There is nothing to catch up.\n");
+	return; // TODO the cold-start case. No need to catch up.
+      }
+
+    constexpr int MAX_PAGE_REQUEST_CNT_AT_ONCE = 10; // Arbitrarily chosen
+    const LOG_PAGEID start_pageid = log_Gl.hdr.append_lsa.pageid;
+    const LOG_PAGEID end_pageid = catchup_lsa.offset == 0 ? catchup_lsa.pageid - 1 : catchup_lsa.pageid;
+    const size_t total_page_count = end_pageid - start_pageid + 1;
+
+    _er_log_debug (ARG_FILE_LINE, "[CATCH-UP] The catch-up pages range from %lld to %lld, count = %lld.\n",
+		   start_pageid, end_pageid, total_page_count);
+
+    const int page_cnt_in_buffer = (int) std::min (total_page_count, (size_t) MAX_PAGE_REQUEST_CNT_AT_ONCE);
+    const size_t buffer_size_in_total = page_cnt_in_buffer * LOG_PAGESIZE + MAX_ALIGNMENT;
+
+    auto log_pgbuf_buffer = std::unique_ptr<char []> { new char[buffer_size_in_total] };
+    char *const aligned_log_pgbuf = PTR_ALIGN (log_pgbuf_buffer.get (), MAX_ALIGNMENT);
+    std::vector<LOG_PAGE *> log_pgptr_vec;
+    char *log_pgptr = aligned_log_pgbuf;
+
+    for (int i = 0; i < page_cnt_in_buffer; i++)
+      {
+	log_pgptr_vec.emplace_back (reinterpret_cast<LOG_PAGE *> (aligned_log_pgbuf + i * LOG_PAGESIZE));
+      }
+
+    LOG_PAGEID request_start_pageid = start_pageid;
+    size_t remaining_page_cnt = total_page_count;
+    while (remaining_page_cnt > 0)
+      {
+	const int request_page_cnt = (int) std::min ((size_t) page_cnt_in_buffer, remaining_page_cnt);
+	// TODO request next log pages asynchronously with std::async
+	const int error_code = m_ps.m_followee_conn->request_log_pages (request_start_pageid, request_page_cnt, log_pgptr_vec);
+	if (error_code != NO_ERROR)
+	  {
+	    /* There are two kinds of erorrs either from client-side or server-side.
```

Comment:
Nit. There are some typos in comments.

Reply:
Thank you. I'll fix them in the following PR.